infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/StreamsSnapshotWriter.java

```diff
@@ -0,0 +1,273 @@
+            UUID shadowID = CorfuRuntime.getStreamID(shadowStream);
+            uuidMap.put(streamID, shadowID);
+            uuidMap.put(shadowID, streamID);
+            streamViewMap.put(streamID, stream);
+            shadowMap.put(shadowID, shadowStream);
+        }
+    }
+
+    /**
+     * clear all tables registered
+     * TODO: replace with stream API
+     */
+    void clearTables() {
+        CorfuStoreMetadata.Timestamp timestamp = persistedWriterMetadata.getTimestamp();
+        long persistSiteConfigID = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID);
+        long persistSnapStart = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapStart);
+        long persitSeqNum = persistedWriterMetadata.query(timestamp, PersistedWriterMetadata.PersistedWriterMetadataType.LastSnapSeqNum);
+
+        //for transfer phase start
+        if (siteConfigID != persistSiteConfigID || srcGlobalSnapshot != persistSnapStart ||
+                (persitSeqNum + 1)!= recvSeq) {
+            log.warn("Skip current siteConfigID " + siteConfigID + " srcGlobalSnapshot " + srcGlobalSnapshot + " currentSeqNum " + recvSeq +
+                    " persistedMetadata " + persistedWriterMetadata.getSiteConfigID() + " startSnapshot " + persistedWriterMetadata.getLastSnapStartTimestamp() +
+                    " lastSnapSeqNum " + persistedWriterMetadata.getLastSnapSeqNum());
+            return;
+        }
+
+
+        TxBuilder txBuilder = persistedWriterMetadata.getTxBuilder();
+        persistedWriterMetadata.appendUpdate(txBuilder, PersistedWriterMetadata.PersistedWriterMetadataType.SiteConfigID, siteConfigID);
+
+
+        for (UUID streamID : streamViewMap.keySet()) {
+            UUID usedStreamID = streamID;
+            if (phase == Phase.TransferPhase) {
+                usedStreamID = uuidMap.get(streamID);
+            }
+
+            SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+            txBuilder.logUpdate(usedStreamID, entry);
+        }
+
+        txBuilder.commit(timestamp);
+    }
+
+    /**
+     * If the metadata has wrong message type or baseSnapshot, throw an exception
+     * @param metadata
+     * @return
+     */
+    void verifyMetadata(LogReplicationEntryMetadata metadata) throws ReplicationWriterException {
+        if (metadata.getMessageMetadataType() != MessageType.SNAPSHOT_MESSAGE ||
+                metadata.getSnapshotTimestamp() != srcGlobalSnapshot) {
+            log.error("snapshot expected {} != recv snapshot {}, metadata {}",
+                    srcGlobalSnapshot, metadata.getSnapshotTimestamp(), metadata);
+            throw new ReplicationWriterException("Message is out of order");
+        }
+    }
+
+    /**
+     * Reset snapshot writer state.
+     * @param snapshot
+     */
+    public void reset(long siteConfigID, long snapshot) {
```

Comment:
are these calls multithreaded. if yes then we need to make sure synchronization is correct.